azurerm/internal/services/consumption/helpers.go

```diff
@@ -0,0 +1,288 @@
+
+	return tags
+}
+
+func ExpandConsumptionBudgetFilter(i []interface{}) *consumption.BudgetFilter {
+	if len(i) == 0 || i[0] == nil {
+		return nil
+	}
+	input := i[0].(map[string]interface{})
+
+	filter := consumption.BudgetFilter{}
+
+	notBlock := input["not"].([]interface{})
+	if len(notBlock) != 0 && notBlock[0] != nil {
+		not := notBlock[0].(map[string]interface{})
+
+		tags := ExpandConsumptionBudgetFilterTag(not["tag"].([]interface{}))
+		dimensions := ExpandConsumptionBudgetFilterDimensions(not["dimension"].([]interface{}))
+
+		if len(dimensions) != 0 {
+			filter.Not = &dimensions[0]
+		} else if len(tags) != 0 {
+			filter.Not = &tags[0]
+		}
+	}
+
+	tags := ExpandConsumptionBudgetFilterTag(input["tag"].(*schema.Set).List())
+	dimensions := ExpandConsumptionBudgetFilterDimensions(input["dimension"].(*schema.Set).List())
+
+	tagsSet := len(tags) > 0
+	dimensionsSet := len(dimensions) > 0
+
+	if dimensionsSet && tagsSet {
+		and := append(dimensions, tags...)
+		filter.And = &and
+	} else {
+		if dimensionsSet {
+			if len(dimensions) > 1 {
+				filter.And = &dimensions
+			} else {
+				filter.Dimensions = dimensions[0].Dimensions
+			}
+		} else if tagsSet {
+			if len(tags) > 1 {
+				filter.And = &tags
+			} else {
+				filter.Tags = tags[0].Tags
+			}
+		}
+	}
+
+	return &filter
+}
+
+func FlattenConsumptionBudgetFilter(input *consumption.BudgetFilter) []interface{} {
+	filter := make([]interface{}, 0)
+
+	if input == nil {
+		return nil
+	}
+
+	dimensions := make([]interface{}, 0)
+	tags := make([]interface{}, 0)
+
+	filterBlock := make(map[string]interface{})
+
+	notBlock := make(map[string]interface{})
+
+	if input.Not != nil {
+		if input.Not.Dimensions != nil {
+			notBlock["dimension"] = []interface{}{FlattenConsumptionBudgetComparisonExpression(input.Not.Dimensions)}
+		}
+
+		if input.Not.Tags != nil {
+			notBlock["tag"] = []interface{}{FlattenConsumptionBudgetComparisonExpression(input.Not.Tags)}
+		}
+
+		if len(notBlock) != 0 {
+			filterBlock["not"] = []interface{}{notBlock}
+		}
+	}
+
+	if input.And != nil {
+		for _, v := range *input.And {
+			if v.Dimensions != nil {
+				dimensions = append(dimensions, FlattenConsumptionBudgetComparisonExpression(v.Dimensions))
+			} else {
+				tags = append(tags, FlattenConsumptionBudgetComparisonExpression(v.Tags))
+			}
+		}
+
+		if len(dimensions) != 0 {
+			filterBlock["dimension"] = dimensions
+		}
+
+		if len(tags) != 0 {
+			filterBlock["tag"] = tags
+		}
+	} else {
+		if input.Tags != nil {
+			filterBlock["tag"] = append(tags, FlattenConsumptionBudgetComparisonExpression(input.Tags))
+		}
+
+		if input.Dimensions != nil {
+			filterBlock["dimension"] = append(dimensions, FlattenConsumptionBudgetComparisonExpression(input.Dimensions))
+		}
+	}
+
+	if len(filterBlock) != 0 {
+		filter = append(filter, filterBlock)
+	}
+
+	return filter
+}
+
+// DiffSuppressFuncs
+func DiffSuppressFuncConsumptionBudgetTimePeriodEndDate(k, old, new string, d *schema.ResourceData) bool {
```

Comment:
I may be mistaken but I thought having `Computed: true` on `end_date` would catch this. Is that not the case?

Reply:
I'll need to double check this as I can't recall now.

Reply:
You were right, I can't recall exactly why I felt this was needed at the time or if it was a user error. Refactored out in https://github.com/terraform-providers/terraform-provider-azurerm/pull/9201/commits/7da9f12cd438f3baf582335314e33e3d9375742c